### DISCUSSION TITLE
Energy measurement: Two registers needed for negative and positive energy integral

### DISCUSSION THREAD

**Discussion by hagenbuch** [arendst/Tasmota#16794] [Q&A] [answered]
I have both a Nous A1 and some Shelly 2.5 devices.

Apparently, negative power (injecting PV into the grid for example) is being indicated correctly in Tasmota in general since 2021 (Yes I know, one is not supposed to inject energy via a standard plug, so let's talk about the Shelly then)

However, as the sign of the momentaneous power may flip (in principle) at every single cycle of the alternating current, I think we need **two energy counters (per channel)** to accumulate positive and negative energy separately. Same would be valid for reactive power BTW.

I have tried to understand the specs and  code of the ADE7953 chip but was only half successful but I don't understand if ESP counts pulses or reads registers from the ADE chip.

To make things right I guess it is necessary that the chip itself holds and provides these sums separately in different registers. Or is the ESP polling the ADE every cycle? In the latter case, Tasmota could total the energy into two own different registers.

However, I guess Tasmota is not yet aware of this disctinction? Or did I overlook it?

More and more folks are connecting their small "balcony style" PV installations to the grid but we might forget that there may be a standstill consumption of the DC/AC converter at very low sun or when not used. I would like to have that separately accounted for, just as the official meters do in the OBIS registers 1.8.0 and 2.8.0 (3.8.0, 4.8.0)

Any ideas?

Andreas

**Comment by joba-1** [comment]
ok, my idea is this: 
while I respect your drive for perfection: have you calculated how much difference it would make, even in worst case real life scenarios (not artificial scenarios like swapping every cycle)? My gut feeling is, it will be a fluctuation below the accuracy that you can achieve, even with careful calibration.
I have such a PV installation, and it reads ~1W consumption at night and with daylight, even on a rainy day, it is above that, so I assume that must already be generated energy.
Also it is not very often where consumption is about as high as production and even then crossings are rare (compared to cycles). 
So to avoid false readings I simply switch off the inverter power at sunset and switch it on at sunrise (adjusted by some empirically found offsets for my specific location).

Regards, Joachim

**Reply by hagenbuch** [reply]
I can see your point and many thanks for your quick reply  but I have bigger installations in mind and in the pipeline, for example with batteries and triphase in mind (Shelly EM3).

For me, perfection is not as important but rather to look up **if** the ADE7953 does already have internal distinct registers for the job, I think we could / should use them.

I'd be also fine with updating the totals every second but even in that case we'd need two counters in the Tasmota memory.

Just imagine Tasmota with a bigger installation on a building and clamps.. Clouds alone might often reverse the current flow to or from the grid.

**Comment by arendst** [comment, marked as answer]
Not every shelly is up to the task of monitoring negative energy values; the hardware does not use the correct features of the ADE7953 and reports random positive or negative values.

Only Shelly EM provides correct negative power values (like the Shelly 3EM using ADE7880). For the latter I introduced the Export Active report.

I will enable this also for the Shelly EM. 

![image](https://user-images.githubusercontent.com/11044339/195597155-46360ed4-8664-44e0-8f51-da59726886d1.png)

**Comment by hagenbuch** [comment]
Oh, fantastic, many thanks!

**Comment by hk2k2** [comment]
Is this also Possible for Sonoff Dual R3 which has BL0939 Chip?
Thanks!

**Comment by 2Fblob** [comment]
I know this is an old thread, but I have just finished adding support to this very same feature on the BL0942 under OpenBecken project.
One problem I have is that there is only one counter, so power is reported correctly, but the totals are displayed as consumption - production, for now.
Im looking into compiling a similar binary to Tasmota - How would I go on about enabling this extended reporting, and what are the variables internally where I should provide the valuea from the BL0942 driver? Thanks!
